### PR TITLE
Make the snippet controller use the DocumentManager and Serializer.

### PR DESCRIPTION
**Pull request by dantleech**
This PR focuses on converting the SnippetController to using the Serializer by way of the DocumentManager.

Subsequently this has affected the SnippetRepository class.

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Resources/config/serializer/Document.BasePageDocument.xml [hunk 1]
```diff
@@ -36,7 +36,9 @@
         <property name="uuid" type="string" serialized-name="id" groups="defaultPage,smallPage,previewPage"/>
         <property name="structureType" type="string" groups="previewPage"/>
         <property name="locale" type="string"/>
+
         <property name="path" type="string"/>
+
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Resources/config/serializer/Document.BasePageDocument.xml]: Revert

File: src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/LocaleSubscriber.php [hunk 2]
```diff
@@ -17,6 +17,7 @@
 use Sulu\Component\Content\Document\LocalizationState;
 use Sulu\Component\DocumentManager\Behavior\Mapping\LocaleBehavior;
 use Sulu\Component\DocumentManager\DocumentRegistry;
+use Sulu\Component\Content\Document\Behavior\ShadowLocaleBehavior;
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/LocaleSubscriber.php]: Revert

File: src/Sulu/Bundle/SnippetBundle/Snippet/SnippetRepository.php [hunk 3]
```diff
@@ -19,11 +19,17 @@
...
 use Sulu\Component\PHPCR\SessionManager\SessionManager;
+use Sulu\Component\DocumentManager\DocumentManager;
+use Sulu\Bundle\SnippetBundle\Document\SnippetDocument;
 
 /**
  * Repository class for snippets.
  *
  * Responsible for retrieving snippets from the content repository
+ *
+ * TODO: Remove SessionManager and ContnetMapper dependencies.
```
Comment (dantleech) [inline src/Sulu/Bundle/SnippetBundle/Snippet/SnippetRepository.php]: typo

File: src/Sulu/Bundle/SnippetBundle/Tests/Functional/Controller/SnippetControllerTest.php [hunk 4]
```diff
@@ -179,7 +180,7 @@ public function testIndex($params, $expectedNbResults)
         $this->assertCount($expectedNbResults, $result['_embedded']['snippets']);
 
         foreach ($result['_embedded']['snippets'] as $snippet) {
-            // check if all snippets have a title, even if it is a ghost page
+            // check if all snippets have a title, even if it is a ghost pagej
```
Comment (dantleech) [inline src/Sulu/Bundle/SnippetBundle/Tests/Functional/Controller/SnippetControllerTest.php]: Revert

**Comment by danrot** [comment]
Can you maybe also implement the `ClassResourceInterface` in the `SnippetController`, so that the name of the methods can be reduced to `getAction` instead of `getSnippetAction` and the same for all other methods?

**Comment by dantleech** [comment]
Update is not working from the web UI.

**Comment by dantleech** [comment]
@danrot really not sure how that works or what the purpose is..

File: src/Sulu/Bundle/DocumentManagerBundle/Bridge/DocumentInspector.php [hunk 5]
```diff
@@ -151,13 +151,28 @@ public function getLocalizationState($document)
...
+        return null;
+    }
+
+    /**
+     * Return true if the document is managed by the document manager.
+     *
+     * @param object $document
+     * @return boolean
+     */
+    public function isManaged($document)
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Bridge/DocumentInspector.php]: Rmeove

File: src/Sulu/Bundle/SnippetBundle/Tests/Functional/BaseFunctionalTestCase.php [hunk 6]
```diff
@@ -41,107 +48,78 @@ public function getKernelConfiguration()
...
+        $page->setShadowLocale('de');
+        $page->getStructure()->bind(array(
+            'hotels' => array(),
+        ));
+
+        $this->manager->persist($page, 'en', array(
+            'path' => '/cmf/sulu_io/contents/hotels',
+        ));
+
+
```
Comment (dantleech) [inline src/Sulu/Bundle/SnippetBundle/Tests/Functional/BaseFunctionalTestCase.php]: EL

**Comment by dantleech** [comment]
@danrot ready

**Comment by danrot** [comment]
@dantleech There is still a failing test.

**Comment by dantleech** [comment]
Tests are passing (I didn't change anything aprt from fixing the genuine failing test).

**Comment by danrot** [comment]
Changelog missing

**Comment by danrot** [comment]
For the `SnippetController` I am missing a test like this: https://github.com/sulu-io/sulu/blob/develop/src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerTest.php#L259

At the moment we can't be sure that the localizations are working as expected. Would you mind adding a test like this?

**Comment by danrot** [comment]
The action `/admin/api/snippet/fields` stopped working, and there also seems to be no test for that. Can you fix that and add a test?

It's caused by the name of the action in the `SnippetController` line 300. The `Snippet` has also to be omitted there.

**Pull request merged**